stablehlo/dialect/TypeInference.cpp

```diff
@@ -1783,6 +1783,105 @@ LogicalResult inferDynamicUpdateSliceOp(
   return success();
 }
 
+// We intend to verify the following properties
+// P1. 1 <= rank <= 3
+// P2. Element types agree with fft_type
+// P3. Operand shape dimensions agree with fft_length for the given fft_type
+LogicalResult inferFftOp(
+    Optional<Location> location, Value operand, bool isRfft, bool isIrfft,
+    DenseIntElementsAttr fftLength,
+    SmallVectorImpl<ShapedTypeComponents>& inferredReturnShapes) {
+  auto fftLengthes = fftLength.getValues<int64_t>();
+  int64_t fftRank = fftLength.size();
+
+  // P1.
+  if (fftRank > 3 || fftRank < 1)
+    return emitOptionalError(location, "rank must be between 1 and 3, but got ",
+                             fftRank, ".");
+
+  // P2. Element type agreement
+  // FFT : C -> C
+  // IFFT : C -> C
+  // RFFT : R -> C
+  // IRFFT : C -> R
+  auto operandType = operand.getType().cast<TensorType>();
+  Type operandElementType = operandType.getElementType();
+  // Check the input element type and infer return element type
+  if (isRfft) {
+    if (!operandElementType.isF32() && !operandElementType.isF64()) {
+      return emitOptionalError(
+          location, "RFFT requires f32 or f64 input type, but is given ",
+          operandElementType, ".");
+    }
+  } else {
+    if (!operandElementType.isa<ComplexType>())
+      return emitOptionalError(
+          location, "FFT/IFFT/IRFFT",
```

Comment:
It's neat that technically the test doesn't need updating, but let's still update it to demonstrate the updated error message.

Reply:
Makes sense, updated.